src/core/apparmor-setup.c

```diff
@@ -0,0 +1,96 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#include <errno.h>
+#if HAVE_APPARMOR
+#include <sys/apparmor.h>
+#endif
+#include <unistd.h>
+
+#include "apparmor-setup.h"
+#include "apparmor-util.h"
+#include "fd-util.h"
+#include "fileio.h"
+#include "log.h"
+#include "macro.h"
+#include "string-util.h"
+#include "strv.h"
+
+#if HAVE_APPARMOR
+DEFINE_TRIVIAL_CLEANUP_FUNC(aa_kernel_interface *, aa_kernel_interface_unref);
+#endif
+
+int mac_apparmor_setup(void) {
+#if HAVE_APPARMOR
+        int r;
+        _cleanup_close_ int fd = -1;
+        _cleanup_(aa_kernel_interface_unrefp) aa_kernel_interface *interface = NULL;
+        const char *current_file;
+        const char *profile_file;
+        _cleanup_free_ char *current_profile = NULL;
+
+        if (!mac_apparmor_use()) {
+                log_debug("AppArmor either not supported by the kernel or disabled.");
+                return 0;
+        }
+
+        /* To enable LSM stacking a patch to the kernel is proposed to create a
+         * per-LSM subdirectory to distinguish between the LSMs. Therefore, we
+         * read the file from the LSM specific directory first and only if that
+         * fails the one from the generic directory.
+         */
+        FOREACH_STRING(current_file, "/proc/self/attr/apparmor/current", "/proc/self/attr/current") {
+                r = read_one_line_file(current_file, &current_profile);
+                if (r > 0)
+                        break;
+        }
+        if (current_profile == NULL) {
+                log_warning_errno(r, "Failed to get the current AppArmor profile of systemd from /proc/self/attr/current and /proc/self/attr/apparmor/current, ignoring: %m");
+                return 0;
+        }
+        if (!streq(current_profile, "unconfined")) {
+                log_debug("Systemd is already confined in an AppArmor profile.");
+                return 0;
+        }
+
+        FOREACH_STRING(profile_file, "/etc/systemd/profile.apparmor", "/usr/lib/systemd/profile.apparmor") {
+                fd = open(profile_file, O_RDONLY | O_CLOEXEC);
+                if (fd < 0) {
+                        if (errno == ENOENT)
+                                continue;
+                        log_error_errno(errno, "Failed to open binary AppArmor profile file %s, ignoring: %m", profile_file);
+                } else
+                        break;
+        }
+        /* Do not log anything if both files do not exist, because we assume the
+         * user does not want to use AppArmor for systemd itself in this case.
+         * Other errors have already been logged above.
+         */
+        if (fd < 0 ) {
+                return 0;
+        }
```

Comment:
as per coding style no {} around single-line if blocks...

shouldn't there at least be debug logging for this?

Reply:
I think, that if none of the possible profile files exists, nothing should be logged. The user most likely does not want to use AppArmor. So we should not pollute the log with information users do not need. If any other error than ENOENT has been encountered while opening one of the profile files, that error will already have been logged within the loop.

Reply:
This should not be a warning. It would be lower.